### PR TITLE
Airbrake Plugin: No API calls made or errors created if there is no project ID

### PR DESCRIPTION
**Pull request by moltenice**
## Hey, I just made a Pull Request!

- If there is no project ID present then no API call is made to Airbrake Backend
- If there is no project ID present then the error API isn't called
- The logic was getting a bit complex so I have refactored it to using enums and doing the complex logic before the rendering of components.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Review by punitcse — APPROVED** [review]
🍰

**Review by LightningStairs — APPROVED** [review]

File: plugins/airbrake/src/api/AirbrakeApi.ts [hunk 1]
```diff
@@ -24,3 +24,9 @@ export const airbrakeApiRef = createApiRef<AirbrakeApi>({
 export interface AirbrakeApi {
   fetchGroups(projectId: string): Promise<Groups>;
 }
+
+export class NoProjectIdError extends Error {
```
Comment (Rugvip) [inline plugins/airbrake/src/api/AirbrakeApi.ts:28]: This often doesn't work as expected, see https://github.com/backstage/backstage/blob/bc0b1f25c6ea4c5ab1d488b6ed7d8e1d96d6a606/packages/errors/src/errors/CustomErrorBase.ts#L54

Not sure about the approach of a custom error here tbh. What do you think about checking whether the `projectId` is available before calling the method instead?

Comment (moltenice) [reply plugins/airbrake/src/api/AirbrakeApi.ts:28]: @Rugvip good shout. I'm keeping this custom error to stick with DRY principles with both the mock and production API. But I've removed conditionals checking the error type to be this custom error and instead just check for `projectId`.

As a result, I've managed to remove some more lines of code through further refactoring 😄 

Let me know what you think!

Comment (Rugvip) [reply plugins/airbrake/src/api/AirbrakeApi.ts:28]: Sorry not quite following. What I'm proposing is that you do something like this:

```ts
const { loading, value, error } = useAsync(
  () => projectId && airbrakeApi.fetchGroups(projectId),
  [airbrakeApi, projectId],
);
```

And then if `!projectId` you just render `<MissingAnnotationEmptyState />` straight away

Comment (moltenice) [reply plugins/airbrake/src/api/AirbrakeApi.ts:28]: So I originally refactored this code because abruptly rendering a component in the logic felt hard to follow.

This is why I introduced the enums here as those are the 4 states that this widget can be in:
https://github.com/backstage/backstage/blob/4239b8dba6c490bc7cc97722e43c9699e3d849bf/plugins/airbrake/src/components/EntityAirbrakeWidget/EntityAirbrakeWidget.tsx#L39-L44

`useAsync` requires me to either return the object from Airbrake or throw an error. I can't return undefined or an empty object (and don't want to make the type support these either). So I believe it does almost what you've proposed except it always either returns a result from Airbrake or throws an error.

Because it always has to either return a result or throw an error, having a conditional check for `projectId` won't have much value, because if `projectId` does not exist I still have to return or throw something.

So what I do is use the logic here:
https://github.com/backstage/backstage/blob/4239b8dba6c490bc7cc97722e43c9699e3d849bf/plugins/airbrake/src/components/EntityAirbrakeWidget/EntityAirbrakeWidget.tsx#L62-L67

To set the `componentState` to the correct one in this case there not being a project ID.

I then still throw an error to satisfy `useAsync`:

https://github.com/backstage/backstage/blob/4239b8dba6c490bc7cc97722e43c9699e3d849bf/plugins/airbrake/src/components/EntityAirbrakeWidget/EntityAirbrakeWidget.tsx#L68

And then print `<MissingAnnotationEmptyState />`:

https://github.com/backstage/backstage/blob/4239b8dba6c490bc7cc97722e43c9699e3d849bf/plugins/airbrake/src/components/EntityAirbrakeWidget/EntityAirbrakeWidget.tsx#L95-L100

Hope that makes more sense :)

File: plugins/airbrake/src/api/AirbrakeApi.ts [hunk 2]
```diff
@@ -24,3 +24,9 @@ export const airbrakeApiRef = createApiRef<AirbrakeApi>({
...
+  constructor() {
```
Comment (Rugvip) [inline plugins/airbrake/src/api/AirbrakeApi.ts:29]: Best use `CustomErrorBase` from `@backstage/errors` to make sure this error behaves as one would expect

**Comment by moltenice** [comment]
@Rugvip Thanks for your suggestions. I think I've addressed them all and the code looks a lot cleaner as a result.

Do take a look again when you get a chance :)

File: plugins/airbrake/src/components/EntityAirbrakeWidget/EntityAirbrakeWidget.tsx [hunk 3]
```diff
@@ -36,59 +37,70 @@ const useStyles = makeStyles<BackstageTheme>(() => ({
...
+    return airbrakeApi.fetchGroups(projectId);
+  }, [airbrakeApi, projectId]);
 
-        {!loading && !projectId && (
-          <MissingAnnotationEmptyState
-            annotation={AIRBRAKE_PROJECT_ID_ANNOTATION}
-          />
-        )}
+  if (loading) {
+    componentState = ComponentState.Loading;
```
Comment (freben) [inline plugins/airbrake/src/components/EntityAirbrakeWidget/EntityAirbrakeWidget.tsx:59]: Why not just do `return <Progress />;` here? And same for the rest of the cases. This enum just makes things more complex and adds indirection, it could be removed.

File: plugins/airbrake/src/components/EntityAirbrakeWidget/EntityAirbrakeWidget.tsx [hunk 4]
```diff
@@ -36,59 +37,70 @@ const useStyles = makeStyles<BackstageTheme>(() => ({
...
-    if (error) {
-      errorApi.post(error);
-    }
-  }, [error, errorApi]);
-
-  if (loading || !projectId || error) {
-    return (
-      <InfoCard title="Airbrake groups" variant="gridItem">
-        {loading && <Progress />}
+  const { loading, value, error } = useAsync(() => {
```
Comment (freben) [inline plugins/airbrake/src/components/EntityAirbrakeWidget/EntityAirbrakeWidget.tsx:54]: Here inside the function you can just do `if (!projectId) { return null; }` or whatever - i disagree with the previous comment, you can absolutely do this and then handle the outcome in the code below, this is all unnecessary. Unless I am gravely mistaken, this whole pull request could be boiled down to roughly four effective lines of code changed if simplified, and turn out to be much more easy to read. :)

File: plugins/airbrake/src/api/ProductionApi.ts [hunk 5]
```diff
@@ -15,13 +15,17 @@
...
 import { Groups } from './airbrakeGroups';
-import { AirbrakeApi } from './AirbrakeApi';
+import { AirbrakeApi, NoProjectIdError } from './AirbrakeApi';
 import { DiscoveryApi } from '@backstage/core-plugin-api';
 
 export class ProductionAirbrakeApi implements AirbrakeApi {
   constructor(private readonly discoveryApi: DiscoveryApi) {}
 
   async fetchGroups(projectId: string): Promise<Groups> {
+    if (!projectId) {
```
Comment (freben) [inline plugins/airbrake/src/api/ProductionApi.ts:25]: see comment about doing this on the calling side instead, i really don't think this api should handle a completely faulty input like this

**Review by freben — COMMENTED** [review]
I still agree with the previous reviewer that this handling of things with an exception hidden inside an api is the wrong way to go

File: plugins/airbrake/src/components/EntityAirbrakeWidget/EntityAirbrakeWidget.tsx [hunk 6]
```diff
@@ -40,55 +41,50 @@ export const EntityAirbrakeWidget = ({ entity }: { entity: Entity }) => {
...
-    () => airbrakeApi.fetchGroups(projectId),
-    [airbrakeApi, projectId],
-  );
-
-  useEffect(() => {
-    if (error) {
-      errorApi.post(error);
+  const { loading, value, error } = useAsync(() => {
+    if (!projectId) {
+      return Promise.resolve(undefined);
```
Comment (freben) [inline plugins/airbrake/src/components/EntityAirbrakeWidget/EntityAirbrakeWidget.tsx:48]: you actually don't need the promise.resolve - but no matter, this works just as well!

**Review by freben — APPROVED** [review]
Let's ship it! :)

File: plugins/airbrake/src/components/EntityAirbrakeWidget/EntityAirbrakeWidget.tsx (hunk 6 above)
Comment (moltenice) [reply plugins/airbrake/src/components/EntityAirbrakeWidget/EntityAirbrakeWidget.tsx:48]: Actually you do need it since the function isn't `async` :)

Whether we should make the function async or not is a whole other question, but since I was calling an async function when calling the API, it made sense to just return that and do away with needing an `async` here.

image

Comment (freben) [reply plugins/airbrake/src/components/EntityAirbrakeWidget/EntityAirbrakeWidget.tsx:48]: ah quite right, i am so used to typing it out async that i missed that :)
